File: javascript/meetup/meetup.js
const DAYS_OF_WEEK = [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday'
]

export const meetup = (year, month, descriptor, day) => {
  let i = 0;
  switch(descriptor) {
      case 'teenth' :
        i = 13;
        break;
      case 'first' :
        i = 1;
        break;
      case 'second' :
        i = 8 ;
        break;
      case 'third' :
        i = 15;
        break;
      case 'fourth' :
        i = 22;
        break;
      case 'last' :
        month = month + 1;
        i = -6;
        break;
      default :
        i = 0;
        break;
  }

  for (let x = i; x <= i + 6; x++) {
    const testDate = new Date(year, (month - 1), x);
      if (testDate.getDay() === DAYS_OF_WEEK.indexOf(day)) {
        return testDate;
      }
  }

}

File: ruby/meetup/meetup.rb
require 'date'

class Meetup

  DAYS_OF_WEEK = {
    sunday: 0,
    monday: 1,
    tuesday: 2,
    wednesday: 3,
    thursday: 4,
    friday: 5,
    saturday: 6
  }.freeze

  START_DAYS = {
    first: 1,
    second: 8,
    teenth: 13,
    third: 15,
    fourth: 22
  }.freeze

  def initialize(month, year)
    @month = month
    @year = year
  end

  attr_reader :month, :year

  def day(day_of_week, occurance)
    date = Date.new(year, month, start_day(occurance))

    6.times do
      break if date.wday == DAYS_OF_WEEK[day_of_week]

      date = date.next_day
    end

    date
  end

  private

  def last_start_day
    Date.new(year, month, 1).next_month - 7
  end

  def start_day(occurance)
    return last_start_day.mday if occurance == :last

    START_DAYS[occurance]
  end
end
